hail/src/main/scala/is/hail/expr/ir/TypeCheck.scala

```diff
@@ -5,26 +5,46 @@ import is.hail.expr.types.virtual._
 import is.hail.utils._
 
 object TypeCheck {
-  def apply(ir: BaseIR): Unit = apply(ir, Env.empty, None)
+  def apply(ir: BaseIR): Unit = apply(ir, BindingEnv.empty)
 
 
-  def apply(ir: BaseIR, env: Env[Type], aggEnv: Option[Env[Type]]): Unit = {
+  def apply(ir: BaseIR, env: BindingEnv[Type]): Unit = {
     try {
       ir match {
-        case ir: IR => _apply(ir, env, aggEnv)
-        case _ =>
-          assert(env.m.isEmpty)
-          assert(aggEnv.isEmpty)
+        case ir: IR => _apply(ir, env)
+        case tir: TableIR => _apply(tir)
+        case mir: MatrixIR => _apply(mir)
+        case bmir: BlockMatrixIR => _apply(bmir)
       }
-
     } catch {
       case e: Throwable => fatal(s"Error while typechecking IR:\n${ Pretty(ir) }", e)
     }
   }
 
-  private def _apply(ir: IR, env: Env[Type], aggEnv: Option[Env[Type]]): Unit = {
-    def check(ir: IR, env: Env[Type] = env, aggEnv: Option[Env[Type]] = aggEnv) {
-      _apply(ir, env, aggEnv)
+  private def _apply(tir: TableIR): Unit = checkChildren(tir)
```

Comment:
I'd just call it `check` or something.  `apply` is syntactic sugar for the apply syntax, if you're not going to use that, better to make the name semantic.